### ISSUE TITLE
Helm doesn't generate kube-prometheus chart properly

### ISSUE THREAD

**Issue by cody-adams** [coreos/prometheus-operator#1288]
**What did you do?**
ran `helm install \
    --name mon \
    --namespace monitoring \
    -f custom-values.yaml \
    coreos/kube-prometheus`
**What did you expect to see?**
A valid config for the Prometheus resource being generated
**What did you see instead? Under which circumstances?**

```
Error: release mon failed: Prometheus.monitoring.coreos.com "mon-prometheus" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"monitoring.coreos.com/v1", "kind":"Prometheus", "metadata":map[string]interface {}{"labels":map[string]interface {}{"chart":"prometheus-0.0.31", "heritage":"Tiller", "prometheus":"mon", "release":"mon", "app":"prometheus"}, "name":"mon-prometheus", "namespace":"monitoring", "creationTimestamp":"2018-04-30T21:32:01Z", "uid":"eb94046c-4cbd-11e8-8ae1-0a58ac1f001d", "selfLink":"", "clusterName":""}, "spec":map[string]interface {}{"alerting":map[string]interface {}{"alertmanagers":[]interface {}{map[string]interface {}{"name":"alertmanager", "namespace":"monitoring", "port":9093, "scheme":"http"}}}, "logLevel":"info", "replicas":1, "routePrefix":"/", "storage":map[string]interface {}{"volumeClaimTemplate":map[string]interface {}{"selector":map[string]interface {}{}, "spec":map[string]interface {}{"accessModes":[]interface {}{"ReadWriteOnce"}, "resources":map[string]interface {}{"requests":map[string]interface {}{"storage":"10Gi"}}, "storageClassName":"default"}}}, "affinity":map[string]interface {}{"podAntiAffinity":map[string]interface {}{"preferredDuringSchedulingIgnoredDuringExecution":[]interface {}{map[string]interface {}{"podAffinityTerm":map[string]interface {}{"labelSelector":map[string]interface {}{"matchLabels":map[string]interface {}{"app":"prometheus", "prometheus":"mon"}}, "topologyKey":"kubernetes.io/hostname"}, "weight":100}}}}, "externalUrl":"URL", "paused":false, "resources":map[string]interface {}{}, "retention":"24h", "ruleSelector":map[string]interface {}{"matchLabels":map[string]interface {}{"prometheus":"mon", "role":"alert-rules"}}, "version":"v2.2.1", "baseImage":"quay.io/prometheus/prometheus"}}: validation failure list:
spec.storage.class in body is required
spec.storage.selector in body is required
spec.storage.resources in body is required
```

**Environment**

* Kubernetes version information:

```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.2", GitCommit:"bdaeafa71f6c7c04636251031f93464384d54963", GitTreeState:"clean", BuildDate:"2017-10-24T19:48:57Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

	Azure Kubernetes Service

* Manifests:

```
# Source: kube-prometheus/charts/prometheus/templates/prometheus.yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app: prometheus
    chart: prometheus-0.0.31
    heritage: Tiller
    prometheus: "mon"
    release: mon
  name: mon-prometheus
spec:
  alerting:
    alertmanagers:
      - name: alertmanager
        namespace: monitoring
        port: 9093
        scheme: http
  baseImage: "quay.io/prometheus/prometheus"
  externalUrl: URL
  paused: false
  replicas: 1
  logLevel:  info
  resources:
    {}
  retention: "24h"
  routePrefix: "/"

  ruleSelector:
    matchLabels:
      role: alert-rules
      prometheus: "mon"
  storage:
    volumeClaimTemplate:
      selector: {}
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        storageClassName: default
  version: "v2.2.1"
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - weight: 100
        podAffinityTerm:
          topologyKey: kubernetes.io/hostname
          labelSelector:
            matchLabels:
              app: prometheus
              prometheus: "mon"
```

Here is the custom values yaml I wrote

```
global:
  rbacEnable: false

alertmanager:
  externalUrl: "URL"
  ingress:
    enabled: true
    fqdn: "URL"
    tls:
      - secretName: SECRET
        hosts:
          - URL
  storageSpec:
    volumeClaimTemplate:
      spec:
        storageClassName: 
        accessModes: ["ReadWriteOnce"]
        resources:
          requests:
            storage: 10Gi

prometheus:
  alertingEndpoints:
    - name: "alertmanager"
      namespace: "monitoring"
      port: 9093
      scheme: http
  ingress:
    enabled: true
    fqdn: "URL"
    tls:
      - secretName: SECRET
        hosts:
          - URL
  storageSpec:
    volumeClaimTemplate:
      spec:
        storageClassName: default
        accessModes: ["ReadWriteOnce"]
        resources:
          requests:
            storage: 10Gi
      selector: {}

grafana:
  auth:
    anonymous:
      enabled: "false"
  adminPassword: "YourPass123#"
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: nginx
      kubernetes.io/tls-acme: "true"
    fqdn: "URL"
    tls:
      - secretName: SECRET
        hosts:
          - URL
  storageSpec:
    class: default
    accessMode: "ReadWriteOnce"
    resources:
      requests:
        storage: 10Gi
```

Is there something I'm missing in the custom values? I basically copied the values.yaml file for the storage class but it never generates it properly.

**Comment by ntfrnzn** [comment]
I believe this is the same issue as https://github.com/coreos/prometheus-operator/issues/1289 and https://github.com/coreos/prometheus-operator/issues/1282 which happens because the selector of the service is defined in the helm template directly, as
```
  selector:
    app: {{ template "prometheus.name" . }}
    prometheus: {{ .Values.prometheusLabelValue | default .Release.Name | quote }}
```
(and defaults to the release name, which is "[user-supplied-name]-prometheus"

but the actual labels of the pod are determined by the prometheus operator and you end up with just prometheus="[user-supplied-name]", that is, just the name of the CRD without the trailing chart name.

So I'm pretty sure that selector template should read
```
  selector:
    app: {{ template "prometheus.name" . }}
    prometheus: {{ template "prometheus.fullname" . }}
```
and simply match the name used for the prometheus CRD.

**Comment by ntfrnzn** [comment]
Note that there's at least one other issue with the labels mismatching -- the various servicemonitors must have a matching `prometheus` label in order to get picked up as targets, and without that they are not polled.  It's unclear which of the recently label changes is incorrect.

I find that version 0.0.50 of the kube-prometheus chart seems to work.  So a helm installation with
```
helm install --name=[foo] \
    --namespace=[whatever]   \
    --version=0.0.50 \
...
    coreos/kube-prometheus
```
is functional with the prometheus-operator chart install.

**Comment by gianrubio** [comment]
Fixed by #1261
